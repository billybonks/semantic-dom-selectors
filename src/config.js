class Config {
  constructor() {
    this.reset();
  }

  reset() {
    this.registeredFinders = [];
    this.defaultFinders = [];
    this.errorLevelOptions = {};
    this.customActors = {
      select: [],
      text: [],
      toggle: [],
      button: [],
    };
    this._rootElement = document.documentElement;
  }

  trim(text) {
    return text.replace(/\n/g, ' ').replace(/\s+/g, ' ').trim();
  }


  get rootElement() {
    return this._rootElement;
  }

  set rootElement(element) {
    if (element instanceof Element) {
      this._rootElement = element;
    } else {
      throw new Error(`rootElement Error: You tried to set root element to ${element}`);
    }
  }

  registerFinder(finder) {
    if (!this.errorLevelOptions[finder.key]) {
      this.errorLevelOptions[finder.key] = 1;
    }
    this.registeredFinders.push(finder);
  }

  registerActor({ type, run }) {
    if (!type) {
      throw new Error('You must specify the type of attribute of actor object');
    }
    if (!Object.keys(this.customActors).includes(type)) {
      throw new Error(`Unkown actor type ${type}`);
    }
    this.customActors[type].push(run);
  }

  setErrorLevels(config) {
    this.errorLevelOptions = config;
  }

  get actors() {
    return this.customActors;
  }

  get finders() {
    return this.registeredFinders.concat(this.defaultFinders);
  }

  get rules() {
    const rules = this.finders.reduce((acc, finder) => {
      const config = this.errorLevelOptions[finder.key];
      if (isNaN(config)) {
        acc[finder.key] = 1;
      } else {
        acc[finder.key] = config;
      }
      return acc;
    }, {});
    return rules;
  }
}

export default new Config();
