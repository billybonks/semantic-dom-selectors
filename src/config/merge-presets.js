import defaultPreset from '../presets/default';
import ariaPreset from '../presets/aria-compliant';

function uniq(array) {
  return Array.from(new Set(array));
}

const presets = {
  default: defaultPreset,
  'aria-compliant': ariaPreset,
};

export default function merge(config) {
  const empty = {
    finders: [],
    actors: [],
    rules: {},
  };

  let preset;
  if (config.preset) {
    preset = presets[config.preset];
    if (!preset) {
      throw new Error('unkown preset set for semantic-dom-selectors configuration');
    }
  } else {
    preset = {};
  }
  const normalizedPreset = Object.assign({}, empty, preset);
  const normalizedConfig = Object.assign({}, empty, config);
  const finders = uniq(normalizedPreset.finders.concat(normalizedConfig.finders));
  const actors = uniq(normalizedPreset.actors.concat(normalizedConfig.actors));
  const rules = Object.assign({}, normalizedPreset.rules, normalizedConfig.rules);
  return {
    finders,
    actors,
    rules,
  };
}
