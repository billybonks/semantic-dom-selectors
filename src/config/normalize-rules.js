const RULE_SEVERITY_STRINGS = ['error', 'warn', 'off'];
const RULE_SEVERITY = RULE_SEVERITY_STRINGS.reduce((map, value, index) => {
  map[value] = index;
  return map;
}, {});
const VALID_SEVERITIES = [0, 1, 2, 'off', 'warn', 'error'];

export default function normalizeRules(rules) {
  return Object.keys(rules).reduce((acc, ruleKey) => {
    let normalizedSeverity;
    const severity = rules[ruleKey];
    if (VALID_SEVERITIES.includes(severity)) {
      if (RULE_SEVERITY_STRINGS.includes(severity)) {
        normalizedSeverity = [RULE_SEVERITY[severity]];
      } else {
        normalizedSeverity = severity;
      }
    } else {
      throw new Error(`Invalid severity ${severity} set for ${ruleKey} please use ${VALID_SEVERITIES}`);
    }
    acc[ruleKey] = normalizedSeverity;
    return acc;
  }, {});
}
