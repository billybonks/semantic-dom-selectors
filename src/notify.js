import config from './config';

export default function(rule,  type, labelText, generateMessage){
  let rules = config.rules;
  let level = rules[rule];
  if(isNaN(level)){
    level = 0
  }
  let humanizedType = `${type.charAt(0).toUpperCase() + type.slice(1)} Control`
  let message = generateMessage ?
    generateMessage(humanizedType, labelText) :
    buildMessage(rule, humanizedType, labelText);

  switch(level){
    case 0:
      throw new Error(message)
    case 1:
      //eslint-disable-next-line no-console
      console.warn(message)
    }
}


function buildMessage(rule, type, labelText){
  switch(rule){
    case 'ambiguousLabel' :
      return  `Multiple ${type} labelled ${labelText} where found`;
    case 'missingObject' :
      return `Could not find ${type} labelled '${labelText}'`;
    default :
      return `Custom rule ${rule} found ${type} labelled '${labelText}'`
  }
}
