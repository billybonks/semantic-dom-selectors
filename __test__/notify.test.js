import notify from '../src/notify';
import config from '../src/config';

function setALevel(rule, level) {
  config.registerFinder({
    key: rule,
  });
  config.errorLevelOptions = {
    [rule]: level,
  };
}

function resetConfig() {
  config.errorLevelOptions = {};
  config.registeredFinders = [];
}

describe('Notification module', () => {
  beforeEach(() => {
    jest.spyOn(console, 'warn');
  });

  afterEach(() => {
    jest.restoreAllMocks();
    resetConfig();
  });

  describe('Defaults', () => {
    test('throws error by default', () => {
      expect(notify).toThrowErrorMatchingSnapshot();
    });

    test('Uses term element by default', () => {
      setALevel('test-rule', 1);
      notify('test-rule');
      expect(console.warn).toHaveBeenCalledWith("Custom rule test-rule found Element labelled 'undefined'");
    });
  });

  describe('Customisng the message', () => {
    beforeEach(() => {
      setALevel('test-rule', 1);
    });

    describe('Specifining type', () => {
      test('Type is uppercased and control is appended', () => {
        notify('test-rule', 'button');
        expect(console.warn).toHaveBeenCalledWith("Custom rule test-rule found Button Control labelled 'undefined'");
      });
    });

    describe('generateMessage', () => {
      test('it uses return of function instead', () => {
        notify('test-rule', 'button', 'blank', () => 'troll');
        expect(console.warn).toHaveBeenCalledWith('troll');
      });
    });

    describe('labelText', () => {
      test('it includes labelText in message as is', () => {
        notify('test-rule', 'button', 'save');
        expect(console.warn).toHaveBeenCalledWith("Custom rule test-rule found Button Control labelled 'save'");
      });
    });
  });

  describe('Notification levels', () => {
    describe('level is 0', () => {
      it('throws an error', () => {
        setALevel('test-rule', 0);
        expect(() => {
          notify('test-rule');
        }).toThrow();
      });
    });

    describe('level is 1', () => {
      it('logs to console.warn', () => {
        setALevel('test-rule', 1);
        notify('test-rule');
        expect(console.warn).toHaveBeenCalled();
      });
    });

    describe('level is 2', () => {
      it('does nothing', () => {
        setALevel('test-rule', 2);
        expect(console.warn).not.toHaveBeenCalled();
      });
    });
  });
});
