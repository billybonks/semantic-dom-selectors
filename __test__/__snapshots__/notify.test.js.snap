// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification module Defaults throws error by default 1`] = `"Custom rule undefined found Element labelled 'undefined'"`;
